Remove necessity for RecordComponent::SCALAR

franzpoeschel opened the pull request:
The last layer in the [openPMD hierarchy](https://openpmd-api.readthedocs.io/en/0.14.3/usage/concepts.html) is used for specifying different dimensions (often: x, y, z) for a dataset. In the openPMD-api:
```cpp
series.iterations[0].meshes["E"]["x"]
series.iterations[0].meshes["E"]["y"]
series.iterations[0].meshes["E"]["z"]

series.iterations[0].particles["e"]["position"]["x"]
series.iterations[0].particles["e"]["position"]["y"]
series.iterations[0].particles["e"]["position"]["z"]
```

However, this last layer is not always needed:

```cpp
series.iterations[0].meshes["e_all_chargeDensity"][RecordComponent::SCALAR]

series.iterations[0].particles["i"]["weighting"][RecordComponent::SCALAR]
```

The purpose of `RecordComponent::SCALAR` in here is to state that the last layer should be skipped. From the view of the type system, this makes sense as `meshes["e_all_chargeDensity"]` and `meshes["e_all_chargeDensity"][RecordComponent::SCALAR]` have different types.
From an API perspective however, this is an unfortunate situation and experience shows that it can be hard to explain to users. It would be better if users could directly write:
```cpp
series.iterations[0].meshes["e_all_chargeDensity"]

series.iterations[0].particles["i"]["weighting"]
```

This feature affects the same interface components and requires many of the same changes as another planned feature: The ability to specify entirely custom datasets, e.g. `series.iterations[0].customDatasets()["my"]["custom"]["hierarchy"]`. Development of this PR should keep this planned feature in mind.

Done so far:
* Make infinite Mesh chains possible
- [x] Maybe undo this again, I think it's not needed
* Make RecordComponent a base class of BaseMesh

TODO:
- [x] Merge #1115 first
- [x] Fix copy and move constructors after deriving virtually from Attributable
- [x] make sure that this condition is still properly checked: `A scalar component can not be contained at " "the same time as one or more regular components.`
- [x] Python bindings
- [ ] Check that API changes in `Attributable::myPath()` go ok with Axel's usage (`SCALAR` is no longer printed out)
- [x] everything else
- [x] Having a look at the output of `openpmd-pipe`, there are still some unneeded attributes written by this (basis is a PIConGPU KelvinHelmholtz dataset):
    ```
    $ colordiff <(bpls -A smol_000500.bp5 | sed -E 's/ *attr//') <(bpls -A smol.bp5 | sed -E 's/ *attr//')
    >   double    /data/500/particles/e/particlePatches/numParticles/unitDimension
    >   double    /data/500/particles/e/particlePatches/numParticlesOffset/unitDimension
    >   double    /data/500/particles/i/particlePatches/numParticles/unitDimension
    >   double    /data/500/particles/i/particlePatches/numParticlesOffset/unitDimension
    ```

## Map of the commits in this PR:

**Auxiliary helpers**
b71f94a36 Add helper: openPMD::auxiliary::overloaded

**Prepare for virtual inheritance of Attributable**
0435622d8 Prepare Attributable for virtual inheritance
e7cd246e4 Fix default constructors/operators of Container and BaseRecordComponent
This PR uses virtual inheritance to solve the diamond problem in order to combine the functionalities of `Container<RecordComponent>` (vector meshes) and `RecordComponent` (scalar meshes) into `BaseRecord`:
![Bildschirmfoto vom 2023-05-05 16-06-06](https://user-images.githubusercontent.com/14241876/236480743-800d9299-6990-44ec-ae74-4f2be2df770d.png)

**Auxiliary helpers**
4afb23c17 Add helper: openPMD::auxiliary::overloaded

**Prepare for virtual inheritance of Attributable**
fc20df1a6 Prepare Attributable for virtual inheritance
dbd07b7ca Fix default constructors/operators of Container and BaseRecordComponent
This PR uses virtual inheritance to solve the diamond problem in order to combine the functionalities of `Container<RecordComponent>` (vector meshes) and `RecordComponent` (scalar meshes) into `BaseRecord`:
![Bildschirmfoto vom 2023-05-05 16-06-06](https://user-images.githubusercontent.com/14241876/236480743-800d9299-6990-44ec-ae74-4f2be2df770d.png)

**State that will be used to track if a BaseRecord is scalar or not**
d6f8057d1 Add m_datasetDefined

**Main implementation part 1: Implement new class structure**
a65eda33f Prepare class structure without applying logic yet

**Main implementation part 2: Implement new logic, no longer use SCALAR**
86d478ebb No longer use map entry SCALAR in application logic

**Follow-up: Remove logic that is no longer needed, adapt tests**
60b063fa8 Remove KEEP_SYNCHRONOUS task
a98eacd37 Adapt Coretests

**Follow-up: Properly apply `Container` interface to `BaseRecord` and some fixes for Container interface itself**
f60bdac5d No virtual methods in Container class
10933c3c3 Fully override Container methods in BaseRecord
908789df3 Adapt Container API to C++17

**Follow-up: Remove SCALAR from `myPath`**
e692ff5ba Adapt myPath() functionality

**Python bindings preparation: Refactoring**
8d3ff4637 Factor out create_and_bind_container() function template
e616c32cd Factor out RecordComponent.__setitem__ and __getitem__
c7928832c Consistently use copy semantics in Python API

**Python bindings: Main implementation**
af7dc7697 Apply new class structure to Python API as well

**Follow-up: Fix up openpmd-pipe (it interacted directly with openPMD types via `isinstance()`, so it needs to be fixed**
641bf9ab0 Adapt openpmd-pipe to new design

**Follow-up: some safeguards**
5bab5d0d7 Safeguard: No scalar and vector components side by side
d53578e40 Avoid object slicing when using records as scalar components

**Examples and documentation**
73a6c69be Remove [SCALAR] from all examples
4202b770d Documentation


Note to myself: For updating the commit map after rebasing:
```bash
cd git-repos/openPMD-api
git checkout topic-remove-scalar-component
xsel -bo | sed "$(paste <(xsel -bo | grep -E '^[[:alnum:]]{9}\>' | awk '{print $1}') <(git log --oneline --graph HEAD dev | head -n 21 | tac | awk '{print $2}') | awk '{printf "s/%s/%s/\n", $1, $2}' | tr '\n' ';')" | xsel -bi
```

ax3l commented:
@franzpoeschel ready for rebase (#1115 is in) :tada:

ax3l commented:
@franzpoeschel please rebase due to `clang-format` changes :)

You might find this helpful: https://github.com/pre-commit/pre-commit/issues/1696#issuecomment-974621448 :)

ax3l commented:
@franzpoeschel this PR seems as it needs to be merged first for a chain of other PRs like the graceful error handling.
Can you please rebase it? :)

franzpoeschel commented:
@ax3l This PR is not necessary for any others, but it brings in a lot of changes in our middle-end logic, so I should rebase it some time next week (and cross fingers that it will be somehow mergeable with other PRs)

franzpoeschel commented:
Ah, I had entered the wrong base PR in #1237

franzpoeschel commented:
Currently following this again on the https://github.com/franzpoeschel/openPMD-api/tree/topic-remove-scalar-component-redo branch.

Instead of managing weird inheritance hierarchies like here, that branch tries out composition over inheritance by inheriting virtually from Attributable. Let's see how that goes.

franzpoeschel commented:
Changed to the new branch now
old branch found at https://github.com/franzpoeschel/openPMD-api/tree/topic-remove-scalar-component-old
Not everything reimplemented yet
Hitting "ready for review", so the CI can share its two cents

franzpoeschel commented:
C++ implementation is now mostly working, Python bindings compile, but fail ~~probably due to this issue https://github.com/pybind/pybind11/discussions/4303~~

It's an existing bug in our codebase that is triggered by this PR. In our container bindings for Python we hand out references to C++ objects:
```cpp
    // keep same policy as Container class: missing keys are created
    cl.def(
        "__getitem__",
        [](Map &m, KeyType const &k) -> MappedType & { return m[k]; },
        // ref + keepalive
        py::return_value_policy::reference_internal);
```

After calling `del series` on the Python side, the referenced object is deallocated (for some reason despite the implied keepalive of `reference_internal`). Deregistering the handle in Python after the referenced object has been destroyed in C++ causes trouble when the referenced object has multiple/virtual inheritance in its class hierarchy.

Solution: We have shared_ptr semantics for our object model anyway, so change the return value policy to `copy`. 
Our current Python binding semantics are a bit weird: `del series` should be replaced by something like `series.close()`, since in a garbage-collected language you cannot really rely on things such as destruction to trigger logic. It's probably possible to reestablish those semantics despite the current troubles, but we might want to take the chance and change them.

UPDATE: Now using copy semantics and propagating keepalive through the object hierarchy. Python bindings now working again.

franzpoeschel commented:
Tests now running green again, Python bindings now work again
Still lots of cleanup to do and other things to merge first, but we're getting there

franzpoeschel committed:
Add helper: openPMD::auxiliary::overloaded

franzpoeschel committed:
Prepare Attributable for virtual inheritance

Use only zero-param constructors to avoid diamond initialization
pitfalls

franzpoeschel committed:
Add m_datasetDefined

See in-code documentation.

franzpoeschel committed:
Prepare class structure without applying logic yet

BaseRecord is now derived by its contained RecordComponent type.
If it is scalar, the idea is that the BaseRecord itself is used as a
RecordComponent, without needing to retrieve the [SCALAR] entry.
No logic implemented yet around this, this just prepares the class
structure.
Notice that this will write some unnecessary attributes since the
RecordComponent types initialize some default attributes upon
construction.

franzpoeschel committed:
No longer use map entry SCALAR in application logic

Not yet supported: Backward compatibility for still allowing legacy
access to scalar entries

franzpoeschel committed:
No virtual methods in Container class

Either this way, or make all of them virtual

franzpoeschel committed:
Fully override Container methods in BaseRecord

Special care for legacy usage of SCALAR constant. Implement iteration
API such that it works for scalar components as well.

franzpoeschel committed:
Adapt myPath() functionality

franzpoeschel committed:
Factor out create_and_bind_container() function template

Will later be called by Record-type classes, too

franzpoeschel committed:
Factor out RecordComponent.__setitem__ and __getitem__

Similarly to the Container API, we will need to apply this to
Record-type classes.
Defining `__setitem__` and `__getitem__` for them is sufficient, as all
other members are inherited from RecordComponent.
`__setitem__` and `__getitem__` need special care, as they are inherited
from Container AND from RecordComponent, so some conflict resolution is
needed.

franzpoeschel committed:
Consistently use copy semantics in Python API

franzpoeschel committed:
Apply new class structure to Python API as well

franzpoeschel committed:
Adapt openpmd-pipe to new design

This somewhat demonstrates that this change is slightly API-breaking.
Since openpmd-pipe acts directly on the class structure via
`instanceof()`, fixes are necessary.

franzpoeschel committed:
Remove KEEP_SYNCHRONOUS task

No longer needed, as one object in the openPMD hierarchy is no longer
represented by possibly multiple Writable objects.

franzpoeschel committed:
Adapt Coretests

franzpoeschel committed:
Adapt Container API to C++17

insert() and reverse iterators

franzpoeschel commented on include/openPMD/backend/BaseRecord.hpp:
```diff
@@ -225,87 +564,346 @@
+     * We skip this check as it changes the runtime of this call from
+     * O(last-first) to O(container().size()).
+     */
+    // for (auto it = this->container().begin(); it != end; ++it)
+    // {
+    //     if (it->first == RecordComponent::SCALAR)
+    //     {
+    //         this->container().erase(it);
+    //         throw error::WrongAPIUsage(detail::NO_SCALAR_INSERT);
+    //     }
+    // }
```
intentional

franzpoeschel commented on include/openPMD/backend/BaseRecord.hpp:
```diff
@@ -225,87 +564,346 @@
+     * We skip this check as it changes the runtime of this call from
+     * O(last-first) to O(container().size()).
+     */
+    // for (auto it = this->container().begin(); it != end; ++it)
+    // {
+    //     if (it->first == RecordComponent::SCALAR)
+    //     {
+    //         this->container().erase(it);
+    //         throw error::WrongAPIUsage(detail::NO_SCALAR_INSERT);
+    //     }
+    // }
```
intentional

franzpoeschel committed:
Fix default constructors/operators of Container and BaseRecordComponent

These derive virtually from Attributable, and this avoids that pitfalls
propagate to user code.

franzpoeschel committed:
Safeguard: No scalar and vector components side by side

"A scalar component can not be contained at the same time as one or more regular components."

franzpoeschel committed:
Remove [SCALAR] from all examples

franzpoeschel committed:
Avoid object slicing when using records as scalar components

franzpoeschel committed:
Documentation

franzpoeschel committed:
Adapt to refactored Python bindings after rebasing

ax3l commented:
Restarting CI for latest version
